src/roiextractors/extractors/memmapextractors/memmapextractors.py

```diff
@@ -64,43 +62,86 @@ def get_num_channels(self):
         """
         imaging = imaging_extractor
-        video_to_save = np.memmap(
-            save_path,
-            shape=(
-                imaging.get_num_frames(),
-                imaging.get_num_channels(),
-                imaging.get_image_size()[0],
-                imaging.get_image_size()[1],
-            ),
-            dtype=imaging.get_dtype(),
-            mode="w+",
-        )
-
-        if verbose:
-            for ch in range(imaging.get_num_channels()):
-                print(f"Saving channel {ch}")
-                for i in tqdm(range(imaging.get_num_frames())):
-                    plane = imaging.get_frames(i, channel=ch)
-                    video_to_save[ch, i] = plane
-        else:
-            for ch in range(imaging.get_num_channels()):
-                for i in range(imaging.get_num_frames()):
-                    plane = imaging.get_frames(i, channel=ch)
-                    video_to_save[ch, i] = plane
+        safety_margin = 0.80  # Accept a file smaller than 80 per cent of available memory
+        file_size_in_bytes = Path(imaging.file_path).stat().st_size
+        available_memory_in_bytes = psutil.virtual_memory().available
+
+        memory_limit = available_memory_in_bytes * safety_margin
+        if file_size_in_bytes < memory_limit and not force_chunk:
+            video_data_to_save = imaging.get_frames()
+            memmap_shape = video_data_to_save.shape
+            video_memmap = np.memmap(
+                save_path,
+                shape=memmap_shape,
+                dtype=imaging.get_dtype(),
+                mode="w+",
+            )
+
+            video_memmap[:] = video_data_to_save
+            video_memmap.flush()
 
-        video_to_save.flush()
+        else:
+            chunk_size_in_bytes = int(available_memory_in_bytes * safety_margin)
+            dtype = imaging.get_dtype()
+            type_size = np.dtype(dtype).itemsize
+
+            n_channels = imaging.get_num_channels()
+            pixels_per_frame = n_channels * np.product(imaging.get_image_size())
+            bytes_per_frame = type_size * pixels_per_frame
+            frames_per_chunk = chunk_size_in_bytes // bytes_per_frame
+
+            num_frames = imaging.get_num_frames()
+            memmap_shape = imaging.video_structure.build_video_shape(n_frames=num_frames)
+
+            iterator = range(0, num_frames, frames_per_chunk)
+            if verbose:
+                iterator = tqdm(iterator, ascii=True, desc="Writing to .dat file")
+
+            for frame in iterator:
+                end_frame = min(frame + frames_per_chunk, num_frames)
+
+                # Get the video chunk
+                video_chunk = imaging.get_video(start_frame=frame, end_frame=end_frame)
+
+                # Load the memmap
+                video_memmap = np.memmap(
+                    save_path,
+                    shape=memmap_shape,
+                    dtype=dtype,
+                    mode="w+",
+                )
```

Comment:
Why can't you just keep the `video_memmap` open the entire time and pull a different frame slice at each iteration?

Reply:
Would that work? I am remember the thing that @bendichter mentioned that `memmaps` tend to keep accessed data in the cache.

Reply:
The writing stages here seem a little odd - I'd refer to https://github.com/SpikeInterface/spikeinterface/blob/master/spikeinterface/core/core_tools.py#L181 (minus the parallelization) for more a typical way of doing it

Reply:
> Would that work? I am remember the thing that @bendichter mentioned that memmaps tend to keep accessed data in the cache.

That was something Ben mentioned that Oliver mentioned but I've not seen it produce that behavior w.r.t. buffering processes in practice.

Reply:
Can you be more specific? At a first glance I see a mechanism of 
1) getting the traces.
2) Put them in the memmap.

Save the re-instantiation and deletion of the memmap at each loop it seems to the same to me. What I am missing?

Reply:
The line I linked: https://github.com/SpikeInterface/spikeinterface/blob/master/spikeinterface/core/core_tools.py#L181 - slice a frame region of the writing memmap, set equal to array extraction of data read from the imaging extractor. That's all you need to iterate, and I'd also be surprised if you need to flush each time.

The part I'm say you can ignore is the executor type setup: https://github.com/SpikeInterface/spikeinterface/blob/master/spikeinterface/core/core_tools.py#L240-L242

Also, I just remembered Ben did a good job with this recent tool: https://github.com/catalystneuro/nwb-conversion-tools/blob/main/src/nwb_conversion_tools/tools/hdmf.py#L7-L23, I'd recommend giving it a try (we'll propagate it up to `hdmf` at some point).

Reply:
> The line I linked: https://github.com/SpikeInterface/spikeinterface/blob/master/spikeinterface/core/core_tools.py#L181 - slice a frame region of the writing memmap, set equal to array extraction of data read from the imaging extractor. That's all you need to iterate, and I'd also be surprised if you need to flush each time.

But the code that I have is the same save flushing and loading the memmap. Do you want those to be out of the loop? is that what are you saying?

Reply:
I'm saying your 13 lines: https://github.com/catalystneuro/roiextractors/pull/122/files#diff-f905195c77c6ce028089c3110e56083aedaf38e2592b2367f1af856d40a38148R134-R147

should be able to be done in one line, as in: https://github.com/SpikeInterface/spikeinterface/blob/master/spikeinterface/core/core_tools.py#L181 and https://github.com/catalystneuro/nwb-conversion-tools/blob/main/src/nwb_conversion_tools/tools/hdmf.py#L23

Reply:
They can't be done like that if you don't know which axis the frames are. 
All the machinery over there is to solve that problem.

Reply:
@CodyCBakerPhD 
The requested refactor is implemented in #125 
Let me know if there is something else left to merge this.